File: LeetCode/5LongestPalindromicSubstring/LongestPalindromicSubstring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        # 2D DP -> Time:O(n^2), Space:O(n^2)
        dp = [[0 for i in range(len(s))] for j in range(len(s))]
        for i in range(len(s)):
            for j in range(len(s)-i):
                if i == 0:
                    dp[j][i+j] = 1
                elif i == 1:
                    dp[j][i+j] = 1 if s[j] == s[i+j] else 0
                elif s[j] == s[j+i] and dp[j+1][i+j-1] == 1:
                        dp[j][i+j] = 1
        for i in range(len(s)):
            for j in range(i):
                if dp[j][len(s)-i+j]:
                    return s[j:len(s)-i+j+1]
        return s[0]

File: LeetCode/5LongestPalindromicSubstring/LongestPalindromicSubstring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        // 2D DP -> Time:O(n^2), Space:O(n^2)
        vector<vector<int>> dp;
        for (int i = 0; i < s.size(); i++) {
            vector<int> temp(s.size(), 0);
            dp.push_back(temp);
        }
        for (int i = 0; i < s.size(); i++) {
            for (int j = 0; j < s.size()-i; j++) {
                cout << j << j+i << endl;
                if (i == 0) {
                    dp[j][j+i] = 1;
                }
                else if (i == 1) {
                    dp[j][j+i] = s[j] == s[j+i] ? 1 : 0;
                }
                else if (s[j] == s[j+i] && dp[j+1][j+i-1] == 1) {
                    dp[j][j+i] = 1;
                }
            }
        }
        for (int i = 0; i < s.size(); i++) {
            for (int j = 0; j < i; j++) {
                if (dp[j][s.size()-i+j]) {
                    return s.substr(j, s.size()-i+1);
                }
            }
        }
        return s.substr(0, 1);
    }
};
